Modules/Common/CardanoLikeNativeTokensModule.php

```diff
@@ -0,0 +1,321 @@
+        return (int)pg_fetch_assoc(pg_query($this->db, 'SELECT max(block_no) FROM block'))['max'];
+    }
+
+    final public function ensure_block($block_id, $break_on_first = false)
+    {
+        $block = pg_fetch_assoc(pg_query_params($this->db, 'SELECT id, hash, time, tx_count FROM block WHERE block_no = $1', [$block_id]));
+        $this->block_hash = substr($block['hash'], 2);
+        $this->block_time = $block['time'];
+        $this->block_db_id = (int)$block['id'];
+        $this->transaction_count = (int)$block['tx_count'];
+
+        // We don't really check if all our nodes have the same block here as we always use one db connection to
+        // retrieve data within a single block anyways, so this one won't be throwing ConsensusException
+    }
+
+    final public function pre_process_block($block_id)
+    {
+        $events = [];
+        $block = [];
+
+        $transactions = pg_fetch_all(pg_query_params($this->db,
+            'SELECT id, hash, out_sum, fee, deposit FROM tx WHERE block_id = $1 ORDER BY block_index', [$this->block_db_id]));
+        // or this can be done via a subquery like this: WHERE block_id = (SELECT id FROM block WHERE block_no = $1),
+        // and $block_id is the param
+
+        if (count($transactions) !== $this->transaction_count)
+            throw new ConsensusException('Transaction count mismatch (orphaned block?)');
+
+        if (!$transactions)
+        {
+            $this->set_return_events([]);
+            $this->set_return_currencies([]);
+            return;
+        }
+
+        $in_query = [];
+
+        foreach ($transactions as $transaction)
+        {
+            $block[($transaction['id'])] = $transaction;
+            $block[($transaction['id'])]['inputs'] = [];
+            $block[($transaction['id'])]['outputs'] = [];
+            $in_query[] = $transaction['id'];
+        }
+
+        $in_query = implode(', ', $in_query);
+
+        $outs = pg_fetch_all(pg_query($this->db,
+            "SELECT tx_out.tx_id, tx_out.address, multi_asset.fingerprint, ma_tx_out.ident AS token, ma_tx_out.quantity
+                    FROM tx_out
+                    JOIN ma_tx_out ON (ma_tx_out.tx_out_id = tx_out.id)
+                    JOIN multi_asset ON ma_tx_out.ident = multi_asset.id
+                    WHERE tx_out.tx_id in ({$in_query})
+                    ORDER BY tx_out.tx_id, tx_out.index"));
+
+        foreach ($outs as $out)
+        {
+            $block[($out['tx_id'])]['outputs'][] = $out;
+        }
+
+        // inputs - due to the way foreign keys are chosen in relevant tables, 3-table-join ends up 8-10 times slower then this:
+
+        $ins_1 = pg_fetch_all(pg_query($this->db,
+            "SELECT tx_in.tx_in_id, tx_out.address, tx_in.id as orderby, tx_out.id as nextkey
+                    FROM tx_in
+                    JOIN tx_out ON (tx_in.tx_out_id = tx_out.tx_id AND tx_in.tx_out_index = tx_out.index)
+                    WHERE tx_in.tx_in_id IN ({$in_query})"));
```

Comment:
No order

Reply:
Please refer to foreach expression at line#152, results of this query are unrolled into a dictionary, during such all order is undone and will simply cause unnecessary computations

Reply:
Ack, the ordering is kept in the code. As mentioned in the comment above `ins_1` the needed composite sql query takes to much time to run and was separated to two queries. This speeds up the data ingestion.